### PR TITLE
mcobbett tags harvested from tests

### PR DESCRIPTION
**Pull request by techcobweb**
# Why ?

See issue [Add test tags to the JSON test report produced by Galasactl #1974](https://github.com/galasa-dev/projectmanagement/issues/1974)

- **test runner harvests testcase tags at runtime**
- **generated extended test on project create from CLI now adds a @Tags example.**

### CODE REVIEW

File: modules/cli/how-to-annotate-test-runs-with-a-tag-cloud.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+You can select a group of tests to run based on many factors like test class name, test package name, bundle name. You can also select a group of tests
+from a test portfolio file based on the tags associated with tests. So, for example, you could choose to launch all the tests which have a certain tag.
+
+## How to inject tags into your test results
+
+There are 2 ways of annotating a test run with tags:
+1. Send a collection of tags to the Galasa Service when the test is launched
+2. Add fixed tags to the test source code using an annotation
+
+We will discuss each of these methods in separate secctions.
```
Comment (eamansour) [inline modules/cli/how-to-annotate-test-runs-with-a-tag-cloud.md]: Typo - `secctions`

File: modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/runner/TagHarvester.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        for(String tag: tagsToAdd ) {
+            this.testStructure.addTag(tag);
+        }
+    }
+
+    private void updateTagsInDss(String runName) {
+
+        // Update the dss with the tags we have collected in the test structure.
+        Set<String> tags = this.testStructure.getTags();
+        String key = "run." + runName + "." + "tags" ;
```
Comment (techcobweb) [inline modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/runner/TagHarvester.java:92]: It's a shame that the IRun which reads these doesn't also offer a write-to-dss to group all the dss-manipulation. Maybe this should be moved.

File: modules/cli/how-to-annotate-test-runs-with-a-tag-cloud.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+```shell
+>galasactl runs submit \
+--class dev.galasa.ivts/dev.galasa.ivts.core.CoreManagerIVT \
+--stream ivts \
+--group myGroup \
+--tags tag1,tag2,tag3
+```
+
+Alternatively, one or more tags can be specified using multiple occurrances of the same `--tags` option:
```
Comment (eamansour) [inline modules/cli/how-to-annotate-test-runs-with-a-tag-cloud.md]: Typo - `occurrances` should be `occurrences`

File: modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/runner/TagHarvester.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        String tagsJson = gson.toJson(tagsJsonArray);
+
+        try {
+            this.dss.put(key, tagsJson);
+        } catch (DynamicStatusStoreException ex) {
+            logger.error("Failed to write tags to dss. Ignoring and running the test anyway.",ex);
+        }
+    }
+
+    private void updateTagsInRas() {
```
Comment (techcobweb) [inline modules/framework/galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/internal/runner/TagHarvester.java:109]: I was a bit worried that this is another write to couchdb... performance...etc. one more write for each test which has tags. It no tags then it won't get written though.

I could have left the 'dirty' TestStructure around until the next testcase state change to avoid this, but if it crashed/died before then you'd not have the tagging information collected in the failed run test structure in RAS...

Let me know if you think I should leave it dirty for couchdb...

**Review by eamansour — APPROVED** [review]

**Comment by techcobweb** [comment]
### **Concise Representation and Reasoning of the File Changes**

The changes in this pull request revolve around adding support for tagging tests in Galasa. A new mechanism (`TagHarvester`) was introduced to extract tags from Java test classes at runtime and integrate them into the test reporting system. The updates include:
- A new class, `TagHarvester`, that processes test annotations and integrates tags into dynamic status stores and result archives.
- Modifications to existing components like `TestRunner` to incorporate the `TagHarvester`.
- Adjustments to templates and example test cases to demonstrate the use of the new `@Tags` annotation.
- Supporting documentation and tests to validate the tagging functionality.
These changes enhance test metadata management, offering better categorization and traceability for test results.

### **Alignment of Pull Request Description with File Changes**

The pull request description mentions the addition of a runtime tag harvesting feature and an example of using the `@Tags` annotation in generated test cases. While it accurately reflects the key aspects of the changes, it lacks details on the implementation scope, such as updates to documentation or the introduction of comprehensive tests. Enhancing the description with these specifics would provide a more complete overview. For example:
- "This PR introduces runtime tag harvesting using a new `TagHarvester` class and integrates it into the `TestRunner`. It also updates CLI templates with `@Tags` examples and improves documentation to guide users on tagging functionality."

### **Complexities or Notable Efforts**

The implementation addresses several complexities:
1. **Edge Case Handling**: The `TagHarvester` ensures robustness by filtering out invalid or duplicate tags. Test cases cover scenarios like empty tags or overlapping tags from multiple sources.
2. **Integration Challenges**: Integrating tags into dynamic status stores (`dss`) and result archives (`ras`) while maintaining backward compatibility required thoughtful design.
3. **Technical Debt Reduction**: By centralizing tag processing in a dedicated class (`TagHarvester`), the codebase gains modularity and maintainability.
Overall, the changes demonstrate a concerted effort to enhance functionality while ensuring reliability and code quality.

File: modules/cli/how-to-annotate-test-runs-with-a-tag-cloud.md (hunk 1 above)
Comment (techcobweb) [reply modules/cli/how-to-annotate-test-runs-with-a-tag-cloud.md]: fixed
